server/src/main/java/org/opensearch/gateway/PrimaryShardBatchAllocator.java

```diff
@@ -99,12 +102,13 @@ public void allocateUnassignedBatch(List<ShardRouting> shardRoutings, RoutingAll
         // only fetch data for eligible shards
         final FetchResult<NodeGatewayStartedShardsBatch> shardsState = fetchData(eligibleShards, inEligibleShards, allocation);
 
+        Set<ShardRouting> batchShardRoutingSet = new HashSet<>(shardRoutings);
         RoutingNodes.UnassignedShards.UnassignedIterator iterator = allocation.routingNodes().unassigned().iterator();
         while (iterator.hasNext()) {
             ShardRouting unassignedShard = iterator.next();
             AllocateUnassignedDecision allocationDecision;
 
-            if (shardRoutings.contains(unassignedShard)) {
+            if (unassignedShard.primary() && batchShardRoutingSet.contains(unassignedShard)) {
```

Comment:
What was preventing the replicas shard to not get assigned earlier. Do we need to add any UTs/ITs for this change?

Reply:
Preventing replica shard assignment was handled already by the contains check. Unassigned replica shard from the routing table will not be from batch PrimaryShardBatchAllocator since this batch contains only the primary unassigned shards.

This check allow us to to skip fast if we come across replica shards in our unassigned shards iterator when allocating primary shards.